test-utils/runtime-tester/src/fuzzing.rs

```diff
@@ -423,34 +488,188 @@ impl Scope {
     }
 
     pub fn deploy_contract(&mut self, receiver_account: &Account, contract_id: usize) {
-        let acc_id = receiver_account.usize_id();
+        let acc_id = self.usize_id(&receiver_account);
         self.accounts[acc_id].deployed_contract = Some(contract_id);
     }
+
+    pub fn add_new_key(
+        &mut self,
+        u: &mut Unstructured,
+        account_id: usize,
+        nonce: Nonce,
+    ) -> Result<AddKeyAction> {
+        let permission = {
+            if u.arbitrary::<bool>()? {
+                AccessKeyPermission::FullAccess
+            } else {
+                AccessKeyPermission::FunctionCall(FunctionCallPermission {
+                    allowance: None,
+                    receiver_id: self.random_account(u)?.id.into(),
+                    method_names: vec![],
+                })
+            }
+        };
+        let signer = InMemorySigner::from_seed(
+            self.accounts[account_id].id.clone(),
+            KeyType::ED25519,
+            format!("test{}.{}", account_id, nonce).as_str(),
+        );
+        self.accounts[account_id].keys.insert(
+            nonce,
+            Key {
+                signer: signer.clone(),
+                access_key: AccessKey { nonce, permission: permission.clone() },
+            },
+        );
+        Ok(AddKeyAction {
+            public_key: signer.public_key,
+            access_key: AccessKey { nonce, permission },
+        })
+    }
+
+    pub fn full_access_signer(
+        &self,
+        u: &mut Unstructured,
+        account: &Account,
+    ) -> Result<InMemorySigner> {
+        let account_idx = self.usize_id(&account);
+        let possible_signers = self.accounts[account_idx].full_access_keys();
+        if possible_signers.is_empty() {
+            // this transaction will be invalid
+            Ok(InMemorySigner::from_seed(
+                self.accounts[account_idx].id.clone(),
+                KeyType::ED25519,
+                self.accounts[account_idx].id.as_ref(),
+            ))
+        } else {
+            Ok(u.choose(&possible_signers)?.clone())
+        }
+    }
+
+    pub fn function_call_signer(
+        &self,
+        u: &mut Unstructured,
+        account: &Account,
+        receiver_id: &String,
+    ) -> Result<InMemorySigner> {
+        let account_idx = self.usize_id(&account);
+        let possible_signers = self.accounts[account_idx].function_call_keys(receiver_id);
+        if possible_signers.is_empty() {
+            // this transaction will be invalid
+            Ok(InMemorySigner::from_seed(
+                self.accounts[account_idx].id.clone(),
+                KeyType::ED25519,
+                self.accounts[account_idx].id.as_ref(),
+            ))
+        } else {
+            Ok(u.choose(&possible_signers)?.clone())
+        }
+    }
+
+    pub fn delete_random_key(
+        &mut self,
+        u: &mut Unstructured,
+        account: &Account,
+    ) -> Result<PublicKey> {
```

Comment:
how do we handle the situation where all keys on an account are deleted? Is that expected or no?

Reply:
This function is only called once and before that call we specifically check if account has any keys.